Mathlib/FieldTheory/PrimitiveElement.lean

```diff
@@ -214,6 +256,157 @@ noncomputable def powerBasisOfFiniteOfSeparable : PowerBasis F E :=
+      exact IntermediateField.coe_smul (⊤ : IntermediateField K E) r x,
+    invFun := fun x => ⟨x, mem_top⟩,
+    left_inv := fun x => by simp only,
+    right_inv := fun x => by simp only,
+  }
+
+lemma finite_intermediateField_of_exists_primitive_element.aux_1
+    (S : Set E) (hprim : adjoin F S = ⊤) (K : IntermediateField F E) :
+    adjoin K S = ⊤ := by
+  apply restrictScalars_injective (K := F) (L' := K) (L := E)
+  rw [restrictScalars_top, ← top_le_iff, ← hprim, adjoin_le_iff,
+    coe_restrictScalars, ← adjoin_le_iff]
+
+lemma finite_intermediateField_of_exists_primitive_element.aux_2
+    (α : E) (hprim : F⟮α⟯ = ⊤) (K K' : IntermediateField F E)
+    (heq : (minpoly K α).map (algebraMap K E) = (minpoly K' α).map (algebraMap K' E)) :
+    K = K ⊓ K' := by
+  set g := (minpoly K α).map (algebraMap K E)
+  set K'' := K ⊓ K'
+  let p : Polynomial K'' := Finset.sum g.support fun n => C ⟨g.coeff n,
+    (mem_inf (S := K) (T := K')).2
+      ⟨by rw [coeff_map]; exact ((minpoly K α).coeff n).property,
+        by rw [heq, coeff_map]; exact ((minpoly K' α).coeff n).property⟩⟩ * X ^ n
+  have hp : p.map (algebraMap K'' E) = g := by
+    rw [← sum_C_mul_X_pow_eq g, Polynomial.sum_def, Polynomial.map_sum]
+    congr
+    funext n
+    rw [Polynomial.map_mul, Polynomial.map_pow, Polynomial.map_X, Polynomial.map_C]
+    congr
+  let q := minpoly K'' α
+  obtain ⟨r, hqdvdp⟩ := show q ∣ p by
+    apply minpoly.dvd
+    rw [aeval_def, eval₂_eq_eval_map, hp, ← eval₂_eq_eval_map, ← aeval_def]
+    exact minpoly.aeval K α
+  have hpne0 : q * r ≠ 0 := by
+    apply_fun Polynomial.map (algebraMap K'' E)
+    rw [← hqdvdp, hp, Polynomial.map_zero,
+      Polynomial.map_ne_zero_iff $ NoZeroSMulDivisors.algebraMap_injective K E]
+    exact minpoly.ne_zero $ isAlgebraic_iff_isIntegral.1 $ isAlgebraic_of_finite K α
+  have hpdeg := IntermediateField.adjoin.finrank (K := K) (x := α) (hx := isIntegral_of_finite K α)
+  rw [Field.finite_intermediateField_of_exists_primitive_element.aux_1 F E _ hprim K,
+    Field.finite_intermediateField_of_exists_primitive_element.aux_finrank F E K,
+    ← Polynomial.natDegree_map_eq_of_injective (NoZeroSMulDivisors.algebraMap_injective K E)
+      (minpoly K α),
+    show (minpoly K α).map (algebraMap K E) = g by rfl,
+    ← hp,
+    Polynomial.natDegree_map_eq_of_injective (NoZeroSMulDivisors.algebraMap_injective K'' E),
+    hqdvdp,
+    Polynomial.natDegree_mul (left_ne_zero_of_mul hpne0) (right_ne_zero_of_mul hpne0),
+    ← IntermediateField.adjoin.finrank (K := K'') (x := α) (hx := isIntegral_of_finite K'' α),
+    Field.finite_intermediateField_of_exists_primitive_element.aux_1 F E _ hprim K'',
+    Field.finite_intermediateField_of_exists_primitive_element.aux_finrank F E K''] at hpdeg
+  replace hpdeg : finrank K E ≥ finrank K'' E := by linarith only [hpdeg, Nat.le_add_right]
+  exact (eq_of_le_of_finrank_le' inf_le_left hpdeg).symm
+
+lemma finite_intermediateField_of_exists_primitive_element.aux_fin
```

Comment:
this lemma seems to be true quite generally, I think it should probably move somewhere where people might find it. It could probably also be generalized a bit to domains at least

Reply:
I suspect that it should be true when `E` is a Noetherian integrally closed domain. But currently let's stick to UFD. Maybe it should be moved to next to the statement that `E[X]` is a UFD whenever `E` is a UFD.

Reply:
Moved to `Polynomial.fintypeSubtypeMonicDvd`.